### PR TITLE
Added the MapRemoteFloret

### PR DESCRIPTION
**Pull request by brototyp**
This PR adds the MapRemoteFloret and thus implements the first step towards #216.

I did not yet implement the XML parsing yet and wanted to open up the discussion on how we should do that.

Simulator Screen Shot - iPod touch (7th generation) - 2021-03-13 at 15 20 31
Simulator Screen Shot - iPod touch (7th generation) - 2021-03-13 at 15 20 34

### CODE REVIEW

File: Cauli/CauliViewController/CauliViewController.swift [hunk 1]
```diff
@@ -60,6 +60,11 @@ internal class CauliViewController: UITableViewController {
         let bundle = Bundle(for: SwitchTableViewCell.self)
         tableView.register(UINib(nibName: SwitchTableViewCell.nibName, bundle: bundle), forCellReuseIdentifier: SwitchTableViewCell.reuseIdentifier)
     }
+    
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        tableView.reloadData()
+    }
```
Comment (brototyp) [inline Cauli/CauliViewController/CauliViewController.swift:67]: This makes sure that the list of all florets is updated once the user goes back to this screen. The description of the `MapRemoteFloret` contains the number of enabled mappings. Without this change, the description will not be updated if the user enables or disables a mapping in the settings of the floret.

File: Cauli/Florets/MapRemote/MappingsListDatasource.swift [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import UIKit
+
+internal class MappingsListDatasource: NSObject, UITableViewDataSource {
```
Comment (lukaswuerzburger) [inline Cauli/Florets/MapRemote/MappingsListDatasource.swift]:
```suggestion
internal class MappingsListDataSource: NSObject, UITableViewDataSource {
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,6 +1,7 @@
 # Changelog
 
 ## Unreleased
+* **feature** Added a `MapRemoteFloret` to simply change urls of requests before they are sent. [#216](https://github.com/cauliframework/cauli/issues/216) by @brototyp
```
Comment (lukaswuerzburger) [inline CHANGELOG.md]:
```suggestion
* **feature** Added a `MapRemoteFloret` that can change urls of requests before they are sent. [#216](https://github.com/cauliframework/cauli/issues/216) by @brototyp
```

File: Cauli/Florets/MapRemote/MapRemoteFloret.swift [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import UIKit
+
+/// The MapRemoteFloret can modify the url of a request before the request is performed.
+/// This is esp. helpful to change the app to use a staging or testing server.
```
Comment (lukaswuerzburger) [inline Cauli/Florets/MapRemote/MapRemoteFloret.swift]: I think this is overly specific
```suggestion
/// The `MapRemoteFloret` can modify the url before the request is performed.
/// This is esp. helpful when using a staging or testing server.
```

File: Cauli/Florets/MapRemote/MapRemoteFloret.swift [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+///
+/// The MapRemoteFloret can only modify the url of a request. If you need to update headers please use the `FindReplaceFloret`.
```
Comment (lukaswuerzburger) [inline Cauli/Florets/MapRemote/MapRemoteFloret.swift]:
```suggestion
/// The `MapRemoteFloret` can only modify the url of a request. If you need to update headers please use the `FindReplaceFloret`.
```

File: Cauli/Florets/MapRemote/MapRemoteFloret.swift [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        get {
+            guard let mappings = userDefaults.array(forKey: "Cauli.MapRemoteFloret.enabledMappings") as? [String] else { return [] }
+            return Set(mappings)
+        }
+        set {
+            userDefaults.setValue(Array(newValue), forKey: "Cauli.MapRemoteFloret.enabledMappings")
+        }
+    }
+
+    /// Instantiates a new MapRemoteFloret instance with an array of mappings.
```
Comment (lukaswuerzburger) [inline Cauli/Florets/MapRemote/MapRemoteFloret.swift]:
```suggestion
    /// Instantiates a new `MapRemoteFloret` instance with an array of mappings.
```

File: Cauli/Florets/MapRemote/Mapping.swift [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import Foundation
+
+/// A mapping is a declaration of a single mapping. All urls that
```
Comment (lukaswuerzburger) [inline Cauli/Florets/MapRemote/Mapping.swift]:
```suggestion
/// A `Mapping` is a declaration of a single mapping. All urls that
```

File: Cauli/Florets/MapRemote/Mapping.swift [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+/// fit the description of the `sourceLocation` will be changed
```
Comment (lukaswuerzburger) [inline Cauli/Florets/MapRemote/Mapping.swift]:
```suggestion
/// match the description of the `sourceLocation` will be changed
```

File: Cauli/Florets/MapRemote/Mapping.swift [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+/// according to the description in the `destinationLocation`.
+/// All nil values are ignored in this case.
```
Comment (lukaswuerzburger) [inline Cauli/Florets/MapRemote/Mapping.swift]:
```suggestion
/// All `nil` values are ignored in this case.
```

File: Cauli/Florets/MapRemote/Mapping.swift [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+/// This mapping will redirect all requests to localhost.
+/// ```swift
+/// Mapping(name: "map local", sourceLocation: MappingLocation(), destinationLocation: MappingLocation(host: "localhost")
+/// ``` 
+public struct Mapping {
+    let name: String
+    let sourceLocation: MappingLocation
+    let destinationLocation: MappingLocation
+
+    /// Initializes a new Mapping.
```
Comment (lukaswuerzburger) [inline Cauli/Florets/MapRemote/Mapping.swift]:
```suggestion
    /// Initializes a new `Mapping`.
```

File: Cauli/Florets/MapRemote/Mapping.swift [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+    /// - Parameters:
+    ///   - name: The name of the Mapping. This is used to uniquely identify this mapping.
```
Comment (lukaswuerzburger) [inline Cauli/Florets/MapRemote/Mapping.swift]:
```suggestion
    ///   - name: The name of the mapping. This is used to uniquely identify this mapping.
```

File: Cauli/Florets/MapRemote/Mapping.swift [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    ///   - sourceLocation: This defines all urls this Mapping should apply to.
```
Comment (lukaswuerzburger) [inline Cauli/Florets/MapRemote/Mapping.swift]:
```suggestion
    ///   - sourceLocation: This defines all urls this mapping should apply to.
```

File: Cauli/Florets/MapRemote/Mapping.swift [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+/// A mapping location describes a given part of a url.
+/// The mapping location can be used to either filter urls, or to apply changes on the url.
+public struct MappingLocation {
+    let `protocol`: Protocol?
+    let host: String?
+    let port: Int?
+    let path: String?
+    let query: String?
+
+    /// Instantiates a new MappingLocation.
```
Comment (lukaswuerzburger) [inline Cauli/Florets/MapRemote/Mapping.swift]:
```suggestion
    /// Instantiates a new `MappingLocation`.
```

File: docs/guides/Florets.md [hunk 14]
```diff
@@ -14,4 +14,8 @@ The [NoCacheFloret](https://cauli.works/docs/Classes/NoCacheFloret.html) modifie
 
 ## FindReplaceFloret
 
-The [FindReplaceFloret](https://cauli.works/docs/Classes/FindReplaceFloret.html) can be used to easily replace any part of a request or response by using a [RecordModifier](https://cauli.works/docs/Classes/FindReplaceFloret/RecordModifier.html)
\ No newline at end of file
+The [FindReplaceFloret](https://cauli.works/docs/Classes/FindReplaceFloret.html) can be used to easily replace any part of a request or response by using a [RecordModifier](https://cauli.works/docs/Classes/FindReplaceFloret/RecordModifier.html)
+
+## MapRemoteFloret
+
+The [MapRemoteFloret](https://cauli.works/docs/Classes/MapRemoteFloret.html) can modify the url of a request before the request is performed. This is esp. helpful to change the app to use a staging or testing server.
```
Comment (lukaswuerzburger) [inline docs/guides/Florets.md]: Same as the source code documentation
```suggestion
The [MapRemoteFloret](https://cauli.works/docs/Classes/MapRemoteFloret.html) can modify the url before the request is performed. This is esp. helpful when using a staging or testing server.
```

**Review by lukaswuerzburger — COMMENTED** [review]
LGTM. Honestly, I really like the code! 
Here are my nitpicks on some wording :D I'm not 100% sure about some, though. I just thought we could simplify it here and there. What do you think?

File: Cauli/Florets/MapRemote/MapRemoteFloret.swift [hunk 15]
```diff
@@ -0,0 +1,100 @@
+//
+//  Copyright (c) 2018 cauli.works
```
Comment (pstued) [inline Cauli/Florets/MapRemote/MapRemoteFloret.swift:2]: Would we still use 2018 or the current year 🙃?

File: Cauli/Florets/MapRemote/MappingsListDatasource.swift (hunk 2 above)
Comment (brototyp) [reply Cauli/Florets/MapRemote/MappingsListDatasource.swift]: Good one

File: Cauli/Florets/MapRemote/MapRemoteFloret.swift (hunk 15 above)
Comment (brototyp) [reply Cauli/Florets/MapRemote/MapRemoteFloret.swift:2]: I am kind of tending towards just removing it completely. I honestly thing it's ineffective. There is no such thing as a legal instance "cauli.works" that could claim the copyright. Even if, it's open source. There are different people involved. 

Should we remove the copyright claim header on all files?

File: Cauli/Florets/MapRemote/MapRemoteFloret.swift [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import UIKit
+
+/// The `MapRemoteFloret` can modify the url before the request is performed.
+/// This is esp. helpful when using a staging or testing server.
+///
+/// The `MapRemoteFloret can only modify the url of a request. If you need to update headers please use the `FindReplaceFloret`.
```
Comment (lukaswuerzburger) [inline Cauli/Florets/MapRemote/MapRemoteFloret.swift]:
```suggestion
/// The `MapRemoteFloret` can only modify the url of a request. If you need to update headers please use the `FindReplaceFloret`.
```

File: Cauli/Florets/MapRemote/MapRemoteFloret.swift [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+        get {
+            guard let mappings = userDefaults.array(forKey: "Cauli.MapRemoteFloret.enabledMappings") as? [String] else { return [] }
+            return Set(mappings)
+        }
+        set {
+            userDefaults.setValue(Array(newValue), forKey: "Cauli.MapRemoteFloret.enabledMappings")
+        }
+    }
+
+    /// Instantiates a new `MapRemoteFloret instance with an array of mappings.
```
Comment (lukaswuerzburger) [inline Cauli/Florets/MapRemote/MapRemoteFloret.swift]:
```suggestion
    /// Instantiates a new `MapRemoteFloret` instance with an array of mappings.
```

**Review by lukaswuerzburger — COMMENTED** [review]
Closing backticks 🤓

**Comment by brototyp** [comment]
> Closing backticks 🤓

Oh man. How embarrassing. Thanks! 🙇

File: Cauli/Florets/MapRemote/MapRemoteFloret.swift (hunk 15 above)
Comment (pstued) [reply Cauli/Florets/MapRemote/MapRemoteFloret.swift:2]: Yep! Let's make this happen.
https://github.com/cauliframework/cauli/issues/224

**Pull request merged**